component/otelcol/processor/discovery/discovery.go

```diff
@@ -0,0 +1,147 @@
+// Package discovery provides an otelcol.processor.discovery component.
+package discovery
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/agent/component"
+	"github.com/grafana/agent/component/discovery"
+	"github.com/grafana/agent/component/otelcol"
+	"github.com/grafana/agent/component/otelcol/internal/fanoutconsumer"
+	"github.com/grafana/agent/component/otelcol/internal/lazyconsumer"
+	"github.com/grafana/agent/pkg/river"
+	promsdconsumer "github.com/grafana/agent/pkg/traces/promsdprocessor/consumer"
+)
+
+func init() {
+	component.Register(component.Registration{
+		Name:    "otelcol.processor.discovery",
+		Args:    Arguments{},
+		Exports: otelcol.ConsumerExports{},
+
+		Build: func(o component.Options, a component.Arguments) (component.Component, error) {
+			return New(o, a.(Arguments))
+		},
+	})
+}
+
+// Arguments configures the otelcol.processor.discovery component.
+type Arguments struct {
+	Targets         []discovery.Target `river:"targets,attr"`
+	OperationType   string             `river:"operation_type,attr,optional"`
+	PodAssociations []string           `river:"pod_associations,attr,optional"`
+
+	// Output configures where to send processed data. Required.
+	Output *otelcol.ConsumerArguments `river:"output,block"`
+}
+
+var (
+	_ river.Defaulter = (*Arguments)(nil)
+	_ river.Validator = (*Arguments)(nil)
+)
+
+// DefaultArguments holds default settings for Arguments.
+var DefaultArguments = Arguments{
+	OperationType: promsdconsumer.OperationTypeUpsert,
+	PodAssociations: []string{
+		promsdconsumer.PodAssociationIPLabel,
+		promsdconsumer.PodAssociationOTelIPLabel,
+		promsdconsumer.PodAssociationk8sIPLabel,
+		promsdconsumer.PodAssociationHostnameLabel,
+		promsdconsumer.PodAssociationConnectionIP,
+	},
+}
+
+// SetToDefault implements river.Defaulter.
+func (args *Arguments) SetToDefault() {
+	*args = DefaultArguments
+}
+
+// Validate implements river.Validator.
+func (args *Arguments) Validate() error {
+	err := promsdconsumer.ValidateOperationType(args.OperationType)
+	if err != nil {
+		return err
+	}
+
+	err = promsdconsumer.ValidatePodAssociations(args.PodAssociations)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Component is the otelcol.exporter.discovery component.
+type Component struct {
+	cfg      Arguments
+	consumer *promsdconsumer.Consumer
+	logger   log.Logger
+}
+
+var _ component.Component = (*Component)(nil)
+
+// New creates a new otelcol.exporter.discovery component.
+func New(o component.Options, c Arguments) (*Component, error) {
+	if c.Output.Logs != nil || c.Output.Metrics != nil {
+		level.Warn(o.Logger).Log("msg", "non-trace input detected; this component only works for traces")
+	}
+
+	nextTraces := fanoutconsumer.Traces(c.Output.Traces)
+	consumer, err := promsdconsumer.NewConsumer(nextTraces, c.OperationType, c.PodAssociations, o.Logger)
```

Comment:
If I update pod associations how do those get updated?

Reply:
[consumer.go](https://github.com/grafana/agent/blob/d53425c45bee6737e65a4daa40ab29bac52fb4f1/pkg/traces/promsdprocessor/consumer/consumer.go) has to be updated, as well as functions listing default values for pod associations, such as:

- [static mode](https://github.com/grafana/agent/blob/d53425c45bee6737e65a4daa40ab29bac52fb4f1/pkg/traces/promsdprocessor/prom_sd_processor.go#L54-L62) defaults
- [flow mode](https://github.com/grafana/agent/blob/d53425c45bee6737e65a4daa40ab29bac52fb4f1/component/otelcol/processor/discovery/discovery.go#L49-L55) defaults
- unit tests, such as [here](https://github.com/grafana/agent/blob/d53425c45bee6737e65a4daa40ab29bac52fb4f1/pkg/traces/promsdprocessor/prom_sd_processor.go#L54-L62) and [here](https://github.com/grafana/agent/blob/d53425c45bee6737e65a4daa40ab29bac52fb4f1/pkg/traces/promsdprocessor/consumer/consumer_test.go#L253-L261)

Reply:
Wow, you're absolutely correct that this wasn't working properly. I think it is fixed now - please take a look?

Reply:
Looks good now!